API: Auth: replace zone contents et al

zeha opened the pull request:
### Short description
Implementation of #8717.

For emptying out secondary zones, sending an empty rrset array should work. 

Also:
- forbid negative TTL in RRsets
- forbid changing RRsets in Consumer zones (deleting them all is still okay)
- reduce code directly in updateDomainSettingsFromDocument, to make clang-tidy happy

### Checklist
<!-- please indicate if any of these things are done/included with this Pull Request. Not all boxes need to be checked for the Pull Request to be accepted -->
I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [x] added or modified unit test(s)

cmouse commented on pdns/ws-auth.cc:
```diff
@@ -1896,10 +1896,76 @@ static void apiServerZoneDetail(HttpRequest* req, HttpResponse* resp) {
+        DNSName qname = apiNameToDNSName(stringFromJson(rrset, "name"));
+        apiCheckQNameAllowedCharacters(qname.toString());
+        QType qtype;
+        qtype = stringFromJson(rrset, "type");
```
why not `QType qtype = stringFromJson(rrset, "type");`

zeha replied:
QType cannot initialize from string:
```
ws-auth.cc:2142:13: error: no viable conversion from 'std::string' (aka 'basic_string<char, char_traits<char>, allocator<char>>') to 'QType'
      QType qtype = stringFromJson(rrset, "type");
```

cmouse replied:
right.

cmouse commented on pdns/ws-auth.cc:
```diff
@@ -1896,10 +1896,76 @@ static void apiServerZoneDetail(HttpRequest* req, HttpResponse* resp) {
+          throw ApiException("RRset "+qname.toString()+" IN "+stringFromJson(rrset, "type")+": unknown type given");
+        }
+        if (rrset["records"].is_array()) {
+          int ttl = intFromJson(rrset, "ttl");
```
ttl could be validated to be positive, or maybe it could be done in checkNewRecords?

cmouse replied:
should 0 be default?

zeha replied:
defaulting the ttl seems wrong to me, nobody has ever asked for it in the other endpoints.

I've reworked *FromJson to add uintFromJson and have that raise.

cmouse commented on regression-tests.api/test_Zones.py:
```diff
@@ -2304,11 +2301,68 @@ def test_put_slave_tsig_key_ids_non_existent(self):
-        self.assertIn('A TSIG key with the name', r.json()['error'])
+        self.put_zone(name, payload, expect_error='A TSIG key with the name')
+
+    def test_zone_replace_rrsets_basic(self):
```
could add test for invalid ttl.

zeha replied:
added

zeha commented:
Rebased to fix conflict with 580f1f76bc6a537cdc83642af4fb7bd20a84262d

zeha commented:
This includes the single commit from #12757 to make the api tests not hang.

zeha commented:
green tests :heart:

Habbie commented:
12757 was merged

zeha commented:
This PR grew a few cleanups prompted by clang-tidy. Rebasing them is inconvenient, so I'd appreciate it if this won't linger around for too long.

mind04 commented:
It would be nice to also add a test to ensure that a broken zone (e.g. CNAME and other records exists) is rejected.

zeha commented:
Rebased and added a (multi-case) test as suggested by @mind04. If there are addtl. test suggestions for invalid combinations it should be easy to add them.

mind04 commented:
Requiring a SOA to clear out a secondary zone is harmful. This will result in a secondary denying everything in the zone until the updated zone is re transferred. Without a SOA queries for the zone are answered with a more harmless "refused" until the zone is re transferred.

Apart from cleaning the caches for secondary zones a call to setStale() is required and will speedup the retransfer.
Primary zones require "unsetting" notified serial or a forced notify.

zeha commented:
Added a commit to allow sending empty rrset for secondary zones. This also ignores a `false` return value from dk.rectifyZone, which seems like the right thing to do (even though it could be communicated better from rectifyZone).

mind04 commented:
Clearing consumer zones is fine, loading them makes no sense and should not be possible, unless you want to do the processing as well.

zeha commented:
> Clearing consumer zones is fine, loading them makes no sense and should not be possible, unless you want to do the processing as well.

It is possible today to create a consumer zone and fill it using the other API operations. This PR does not change this.

If consumer zones should be restricted in API, that should go into another PR. Is there a ticket describing the valid and invalid usecases?

zeha commented:
> Allowing records in a consumer zones in in the rest of the api is a bug.
> 
> You should not allow replacing consumer zones unless you want to do the full processing of the new catalog (please don't ;) ). 

Well, I can add some code to forbid changing consumer zones here in this new API. For the existing API surface, thats really out of scope.

> The same applies btw to presigned zones since rectifyZone() is incompatible with presigned zones. rectifyZone() is rectifying according to the PowerDNS rules. A presigned zone does not necessarily follow those rules.  This is why the presigned check is present in rectyfyZone() so please do not ignore the return value.

There's already a check before `dk.rectifyZone` to skip presigned zones. Don't know what the plan is for API-RECTIFY=1 + presigned. Thats probably a misconfiguration.

> Clearing consumer and presigned zones is the only valid option.

It would be really good to have a doc somewhere describing what can be done to what zone type and what is supported. Right now these rules do not seem to be enforced consistently and in some places not at all.

kpfleming commented:
> It would be really good to have a doc somewhere describing what can be done to what zone type and what is supported. Right now these rules do not seem to be enforced consistently and in some places not at all.

Next week I'm going to be working a docs PR to improve the "Modes of Operation" page, and as part of that I'm also planning on a "Zone Kinds" page, since there is a lot of interaction between zone kinds and operation modes.

zeha commented:
I've added additional restrictions for Consumer zones on the existing API surface.

Habbie reviewed (approved):
Wonderful cleanup in the process, thanks. I see nothing obviously wrong, and I get the impression that concerns from @mind04 were addressed.

@mind04 any comment?
